org/mozilla/jss/tests/TestCmac.java

```diff
@@ -0,0 +1,89 @@
+
+import javax.crypto.*;
+import javax.crypto.spec.*;
+
+import org.mozilla.jss.*;
+import org.mozilla.jss.crypto.*;
+import org.mozilla.jss.util.*;
+
+public class TestCmac {
+    private static final byte[] NIST_128 = Base64.getDecoder().decode("K34VFiiu0qar9xWICc9PPA==");
+    private static final byte[] NIST_192 = Base64.getDecoder().decode("jnOw99oOZFLIEPMrgJB55WL46tJSLGt7");
+    private static final byte[] NIST_256 = Base64.getDecoder().decode("YD3rEBXKcb4rc67whX13gR81LAc7YQjXLZgQowkU3/Q=");
+
+    public static void main(String[] args) throws Exception {
+        CryptoManager.initialize(args[0]);
+        CryptoManager cm = CryptoManager.getInstance();
+        CryptoToken tok = cm.getInternalKeyStorageToken();
+        PasswordCallback cb = new FilePasswordCallback(args[1]);
+        tok.login(cb);
+
+        testNISTExamples();
+    }
+
+    /*
+     * The following test vectors come from NIST's Examples with Intermediate
+     * Values page:
+     * https://csrc.nist.gov/projects/cryptographic-standards-and-guidelines/example-values
+     *
+     * These are the same vectors utilized by NSS in:
+     * gtests/freebl_gtest/cmac_unittests.cc
+     *
+     * These same vectors are also found in FRC 4493, Section 4.
+     */
+    public static void testNISTExamples() throws Exception {
+        byte[] all_input = Base64.getDecoder().decode("a8G+4i5An5bpPX4Rc5MXKq4tilceA6ycnrdvrEWvjlEwyBxGo1zkEeX7wRkaClLv9p8kRd9PmxetK0F75mw3EA==");
+        int[] input_lengths = new int[] { 0, 16, 20, 64};
+
+        byte[][] all_expected = new byte[][] {
+            Base64.getDecoder().decode("ux1pKelZNyh/o30Sm3VnRg=="),
+            Base64.getDecoder().decode("BwoWtGtNQUT3m92d0EoofA=="),
+            Base64.getDecoder().decode("fYVEnqbqGcgjp794g3363g=="),
+            Base64.getDecoder().decode("UfC+v347nZL8SXQXeTY8/g=="),
+            Base64.getDecoder().decode("0X3fRq2qzeUxysSD3nqTZw=="),
+            Base64.getDecoder().decode("npmnvzHnEJAGYvZeYXxRhA=="),
+            Base64.getDecoder().decode("PXXBlO2WBwREqfp+x0Ds+A=="),
+            Base64.getDecoder().decode("odXfDu15D3lNd1iWWfOaEQ=="),
+            Base64.getDecoder().decode("Aoli9ht7+J78a1UfRmfZgw=="),
+            Base64.getDecoder().decode("KKcCP0Uuj4K9S/KNjDfDXA=="),
+            Base64.getDecoder().decode("FWcn3Ah4lEoCPB/gO61tkw=="),
+            Base64.getDecoder().decode("4ZkhkFSfbtVpaiwFbDFUEA==")
+        };
+
+        for (int i = 0; i < all_expected.length; i++) {
+            byte[] key = getKey(i);
+            byte[] input = Arrays.copyOf(all_input, input_lengths[i % input_lengths.length]);
+            byte[] expected = all_expected[i];
+
+            testCMAC(key, input, expected);
+        }
+    }
+
+    public static byte[] getKey(int index) {
+        if (index < 4) {
+            return NIST_128;
+        } else if (index < 8) {
+            return NIST_192;
+        } else if (index < 12) {
+            return NIST_256;
+        }
+
+        return null;
+    }
+
```

Comment:
Question:

In TPS/TKS we start with a symmetric key such as the master key and then send that down into the routine that calculates the CMAC. The key will be on the hsm possibly as a SymmetricKey object variable. How do we use this from JSS from that starting point? The test examples start from byte array keys, which is what you expect in the HMAC stuff. 

Would it be similar to what I found in JSS here:

// perform the digesting
        JSSMessageDigest digest = token.getDigestContext(HMACAlgorithm.SHA1);
        digest.initHMAC(key);
        byte[] digestBytes = digest.digest(toBeMACed);
Or is there a way to use the nice provider support we have here starting with a sym key object?

Reply:
So there's two interfaces and a class in question here... and a pointer to the PKI code might be nice. But I'll assume the following is correct. :)

Classes/Interfaces in question:

 - [`javax.crypto.SecretKey`](https://docs.oracle.com/javase/8/docs/api/javax/crypto/SecretKey.html) an interface,
 - [`org.mozilla.jss.crypto.SymmetricKey`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/crypto/SymmetricKey.html) an interface which extends `SecretKey`,
 - [`org.mozilla.jss.crypto.SecretKeyFacade`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/crypto/SecretKeyFacade.html) a class which implements the above `SecretKey` interface via wrapping a `SymmetricKey`-implementing instance,

Everything that JSS uses has to be an instance of `SecretKeyFacade`. Internally, the `key` member (of interface `SymmetricKey`) is usually of type [`org.mozilla.jss.pkcs11.PK11SymKey`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/pkcs11/PK11SymKey.html). The latter includes additional information like algorithm, and a NSS `PK11SymKey` pointer. So when you say you have a `SymmetricKey` instance, you usually actually have a `PK11SymKey` instance. But anyways...

To go from `SymmetricKey` to something usable by the JSS Provider (and, not just the underlying crypto), your easiest option is to do something like:

```java
javax.crypto.SecretKey key = new org.mozilla.jss.crypto.SecretKeyFacade(mySymmetricKey);
```

That's really what [`JSSSecretKeyFactorySpi`](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/provider/javax/crypto/JSSSecretKeyFactorySpi.java) does [under the covers](https://github.com/dogtagpki/jss/blob/master/org/mozilla/jss/provider/javax/crypto/JSSSecretKeyFactorySpi.java#L61).

Note that `JSSSecretKeyFactorySpi` implements the [`javax.crypto.SecretKeyFactorySpi`](https://docs.oracle.com/javase/8/docs/api/javax/crypto/SecretKeyFactorySpi.html) SPI class, so it can eventually be called from the Java Provider [`javax.crypto.SecretKeyFactory`](https://docs.oracle.com/javase/8/docs/api/javax/crypto/SecretKeyFactory.html) member (when the provider is `Mozilla-JSS`).

That brings us full-circle: if you have bytes, you can use the SecretKeyFactory with a correct `spec` class to get a `SecretKey` instance to use, otherwise, if you already have a PKCS#11-backed key from JSS, you can just use `SecretKeyFacade` to wrap it.

Reply:
See #271. I think we can support just passing SymmetricKey directly here and be better off.

Reply:
That is the info I wanted! Just double checking I can do what we need :)